query-engine/core/src/interpreter/query_interpreters/read.rs

```diff
@@ -120,18 +128,26 @@ fn read_many(
             (scalars, aggregation_rows)
         };
 
-        let nested: Vec<QueryResult> = process_nested(tx, query.nested, Some(&scalars)).await?;
-
-        Ok(RecordSelection {
-            name: query.name,
-            fields: query.selection_order,
-            scalars,
-            nested,
-            query_arguments: query.args,
-            model: query.model,
-            aggregation_rows,
+        if scalars.records.is_empty()
+            && query
+                .options
+                .iter()
+                .any(|option| matches!(option, QueryOption::ThrowOnEmpty))
+        {
+            Err(QueryGraphBuilderError::RecordNotFound("Record not found".to_owned()).into())
+        } else {
+            let nested: Vec<QueryResult> = process_nested(tx, query.nested, Some(&scalars)).await?;
+            Ok(RecordSelection {
+                name: query.name,
+                fields: query.selection_order,
+                scalars,
+                nested,
+                query_arguments: query.args,
+                model: query.model,
+                aggregation_rows,
+            }
+            .into())
```

Comment:
TODO: refactor this to reduce cyclomatic complexity

Reply:
Done in https://github.com/prisma/prisma-engines/pull/3356/commits/03e20050e0a09c718e33b8e6281e81d5058b8352